library/src/main/java/org/apache/apex/malhar/lib/fs/s3output/S3FileMerger.java

```diff
@@ -0,0 +1,301 @@
+  {
+    @Override
+    public void process(S3BlockUpload.UploadBlockMetadata tuple)
+    {
+      processUploadBlock(tuple);
+    }
+  };
+
+  /**
+   * Process to merge the uploaded block into a file.
+   * @param tuple uploaded block meta data
+   */
+  protected void processUploadBlock(S3BlockUpload.UploadBlockMetadata tuple)
+  {
+    List<PartETag> listOfUploads = uploadParts.get(tuple.getKeyName());
+    if (listOfUploads == null) {
+      listOfUploads = new ArrayList<>();
+      uploadParts.put(tuple.getKeyName(), listOfUploads);
+    }
+    listOfUploads.add(tuple.getPartETag());
+    if (fileMetadatas.get(tuple.getKeyName()) != null) {
+      verifyAndEmitFileMerge(tuple.getKeyName());
+    }
+  }
+
+  /**
+   * Input port to receive UploadFileMetadata
+   */
+  public final transient DefaultInputPort<S3InitiateFileUpload.UploadFileMetadata> filesMetadataInput = new DefaultInputPort<S3InitiateFileUpload.UploadFileMetadata>()
+  {
+    @Override
+    public void process(S3InitiateFileUpload.UploadFileMetadata tuple)
+    {
+      processFileMetadata(tuple);
+    }
+  };
+
+  /**
+   * Process to merge the uploaded blocks for the given file metadata.
+   * @param tuple file metadata
+   */
+  protected void processFileMetadata(S3InitiateFileUpload.UploadFileMetadata tuple)
+  {
+    String keyName = tuple.getKeyName();
+    fileMetadatas.put(keyName, tuple);
+    if (uploadParts.get(keyName) != null) {
+      verifyAndEmitFileMerge(keyName);
+    }
+  }
+
+  @Override
+  public void setup(Context.OperatorContext context)
+  {
+    windowDataManager.setup(context);
+    s3Client = createClient();
+  }
+
+  /**
+   * Create AmazonS3 client using AWS credentials
+   * @return AmazonS3
+   */
+  protected AmazonS3 createClient()
+  {
+    AmazonS3 client = new AmazonS3Client(new BasicAWSCredentials(accessKey, secretAccessKey));
+    if (endPoint != null) {
+      client.setEndpoint(endPoint);
+    }
+    return client;
+  }
+
+
+  @Override
+  public void beginWindow(long windowId)
+  {
+    currentWindowId = windowId;
+  }
+
+  @Override
+  public void endWindow()
+  {
+    if (uploadedFiles.size() > 0) {
+      for (String keyName: uploadedFiles) {
+        uploadParts.remove(keyName);
+        fileMetadatas.remove(keyName);
+      }
+      uploadedFiles.clear();
+    }
+    if (currentWindowId > windowDataManager.getLargestCompletedWindow()) {
+      try {
+        windowDataManager.save("Uploaded Files", currentWindowId);
+      } catch (IOException e) {
+        throw new RuntimeException("Unable to save recovery", e);
+      }
+    }
+  }
+
+  @Override
+  public void teardown()
+  {
+    windowDataManager.teardown();
+  }
+
+  /**
+   * Send the CompleteMultipartUploadRequest to S3 if all the blocks of a file are uploaded into S3.
+   * @param keyName file to upload into S3
+   */
+  private void verifyAndEmitFileMerge(String keyName)
+  {
+    if (currentWindowId <= windowDataManager.getLargestCompletedWindow()) {
```

Comment:
No. This can't be moved to  processFileMetadata and processBlockUpload methods. Because I am not saving the uploadParts and fileMetadatas in WAL.